pkg/machinery/client/connection.go

```diff
@@ -0,0 +1,192 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	clientconfig "github.com/talos-systems/talos/pkg/machinery/client/config"
+	"github.com/talos-systems/talos/pkg/machinery/client/resolver"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+	"github.com/talos-systems/talos/pkg/machinery/generic/slices"
+)
+
+// getConn creates new gRPC connection.
+func (c *Client) getConn(ctx context.Context, opts ...grpc.DialOption) (*grpcConnectionWrapper, error) {
+	endpoints := c.GetEndpoints()
+
+	target := c.getTarget(
+		resolver.EnsureEndpointsHavePorts(
+			reduceURLsToAddresses(endpoints),
+			constants.ApidPort),
+	)
+
+	dialOpts := []grpc.DialOption(nil)
+
+	dialOpts = append(dialOpts, c.options.grpcDialOptions...)
+
+	dialOpts = append(dialOpts, opts...)
+
+	if c.options.unixSocketPath != "" {
+		conn, err := grpc.DialContext(ctx, target, dialOpts...)
+
+		return newGRPCConnectionWrapper(c.GetClusterName(), conn), err
+	}
+
+	tlsConfig := c.options.tlsConfig
+
+	if tlsConfig != nil {
+		return c.makeConnection(ctx, target, credentials.NewTLS(tlsConfig), dialOpts)
+	}
+
+	if err := c.resolveConfigContext(); err != nil {
+		return nil, fmt.Errorf("failed to resolve configuration context: %w", err)
+	}
+
+	basicAuth := c.options.configContext.Auth.Basic
+	if basicAuth != nil {
+		dialOpts = append(dialOpts, WithGRPCBasicAuth(basicAuth.Username, basicAuth.Password))
+	}
+
+	creds, err := buildCredentials(c.options.configContext, endpoints)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.makeConnection(ctx, target, creds, dialOpts)
+}
+
+func buildTLSConfig(configContext *clientconfig.Context) (*tls.Config, error) {
+	tlsConfig := &tls.Config{}
+
+	caBytes, err := getCA(configContext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get CA: %w", err)
+	}
+
+	if len(caBytes) > 0 {
+		tlsConfig.RootCAs = x509.NewCertPool()
+		tlsConfig.ClientCAs = x509.NewCertPool()
+
+		if ok := tlsConfig.ClientCAs.AppendCertsFromPEM(caBytes); !ok {
```

Comment:
`.ClientAuth` and `.ClientCAs` are fields which are used by the server-side `tls.Config`, they shouldn't be set here: https://pkg.go.dev/crypto/tls#Config